File: plugins/sulis-execution/sdk/python/sulis_execution/resources/wp.py
"""WP resource — wraps `wpx-wp` subcommands."""
from __future__ import annotations

from sulis_execution._helpers import _result_payload
from sulis_execution.transport import (
    AsyncSubprocessTransport,
    SubprocessTransport,
    TransportConfig,
)
from sulis_execution.types import WpAppendEvidenceResult, WpReadFrontmatterResult

BINARY = "wpx-wp"


def _common(config: TransportConfig) -> dict:
    return {"project": config.project, "repo_root": str(config.repo_root)}


class WpResource:
    def __init__(
        self, transport: SubprocessTransport, config: TransportConfig
    ) -> None:
        self._transport = transport
        self._config = config

    def read_frontmatter(self, *, wp: str, field: str) -> WpReadFrontmatterResult:
        params = {**_common(self._config), "wp": wp, "field": field}
        envelope = self._transport.invoke(BINARY, "read-frontmatter", params)
        return WpReadFrontmatterResult.model_validate(_result_payload(envelope))

    def append_evidence(
        self, *, wp: str, evidence_json: str
    ) -> WpAppendEvidenceResult:
        params = {**_common(self._config), "wp": wp, "evidence_json": evidence_json}
        envelope = self._transport.invoke(BINARY, "append-evidence", params)
        return WpAppendEvidenceResult.model_validate(_result_payload(envelope))


class AsyncWpResource:
    def __init__(
        self, transport: AsyncSubprocessTransport, config: TransportConfig
    ) -> None:
        self._transport = transport
        self._config = config

    async def read_frontmatter(
        self, *, wp: str, field: str
    ) -> WpReadFrontmatterResult:
        params = {**_common(self._config), "wp": wp, "field": field}
        envelope = await self._transport.invoke(BINARY, "read-frontmatter", params)
        return WpReadFrontmatterResult.model_validate(_result_payload(envelope))

    async def append_evidence(
        self, *, wp: str, evidence_json: str
    ) -> WpAppendEvidenceResult:
        params = {**_common(self._config), "wp": wp, "evidence_json": evidence_json}
        envelope = await self._transport.invoke(BINARY, "append-evidence", params)
        return WpAppendEvidenceResult.model_validate(_result_payload(envelope))

File: plugins/sulis-execution/sdk/typescript/src/resources/wp.ts
import { resultPayload } from '../helpers.js';
import type {
  AsyncSubprocessTransport,
  SubprocessTransport,
  TransportConfig,
} from '../transport.js';
import type {
  WpAppendEvidenceResult,
  WpReadFrontmatterResult,
} from '../types.js';

const BINARY = 'wpx-wp';

function common(config: TransportConfig): Record<string, unknown> {
  return { project: config.project, repo_root: config.repoRoot };
}

export class WpResource {
  constructor(
    private readonly transport: SubprocessTransport,
    private readonly config: TransportConfig,
  ) {}

  readFrontmatter(params: { wp: string; field: string }): WpReadFrontmatterResult {
    return resultPayload(
      this.transport.invoke(BINARY, 'read-frontmatter', {
        ...common(this.config),
        ...params,
      }),
    ) as unknown as WpReadFrontmatterResult;
  }

  appendEvidence(params: {
    wp: string;
    evidence_json: string;
  }): WpAppendEvidenceResult {
    return resultPayload(
      this.transport.invoke(BINARY, 'append-evidence', {
        ...common(this.config),
        ...params,
      }),
    ) as unknown as WpAppendEvidenceResult;
  }
}

export class AsyncWpResource {
  constructor(
    private readonly transport: AsyncSubprocessTransport,
    private readonly config: TransportConfig,
  ) {}

  async readFrontmatter(params: {
    wp: string;
    field: string;
  }): Promise<WpReadFrontmatterResult> {
    return resultPayload(
      await this.transport.invoke(BINARY, 'read-frontmatter', {
        ...common(this.config),
        ...params,
      }),
    ) as unknown as WpReadFrontmatterResult;
  }

  async appendEvidence(params: {
    wp: string;
    evidence_json: string;
  }): Promise<WpAppendEvidenceResult> {
    return resultPayload(
      await this.transport.invoke(BINARY, 'append-evidence', {
        ...common(this.config),
        ...params,
      }),
    ) as unknown as WpAppendEvidenceResult;
  }
}
